e2e_example/test/common/utils.dart

```diff
@@ -0,0 +1,178 @@
+///
+/// For debugging purposes you can enable printing of the build script output by
+/// setting [verbose] to `true`.
+Future<Null> startServer(
+    {bool ensureCleanBuild, bool verbose, List<Function> extraExpects}) async {
+  ensureCleanBuild ??= true;
+  verbose ??= false;
+  extraExpects ??= [];
+
+  // Make sure this is a clean build
+  if (ensureCleanBuild && await _toolDir.exists()) {
+    await _toolDir.delete(recursive: true);
+  }
+
+  _process = await Process.start('dart', ['tool/build.dart']);
+  _stdOutLines = _process.stdout
+      .transform(UTF8.decoder)
+      .transform(const LineSplitter())
+      .asBroadcastStream();
+
+  _stdErrLines = _process.stderr
+      .transform(UTF8.decoder)
+      .transform(const LineSplitter())
+      .asBroadcastStream();
+
+  if (verbose) {
+    _stdOutLines.listen((line) => print('StdOut: $line'));
+    _stdErrLines.listen((line) => print('StdErr: $line'));
+  }
+
+  await Future.wait(extraExpects.map((cb) async => await cb()));
+  await nextSuccessfulBuild;
+}
+
+/// Kills the current build script.
+///
+/// By default this will clean up the `.dart_tool` directory, but you can
+/// disable this by setting [cleanUp] to `false`.
+Future<Null> stopServer({bool cleanUp}) async {
+  cleanUp ??= true;
+  expect(_process.kill(), true);
+  await _process.exitCode;
+  _process = null;
+
+  if (cleanUp) await _toolDir.delete(recursive: true);
+}
+
+/// Checks whether the current git client is "clean" (no pending changes) for
+/// this package directory.
+bool _gitIsClean() {
+  var gitStatus = Process.runSync('git', ['status', '.']).stdout as String;
+  return gitStatus.contains('nothing to commit, working tree clean');
+}
+
+/// Ensures that the current directory is a "clean" git client (no pending
+/// changes).
+///
+/// This also calls `addTearDown` with a method that will reset the current git
+/// state for this directory after running the test.
+void ensureCleanGitClient() {
+  var gitWasClean = _gitIsClean();
+  expect(gitWasClean, isTrue,
+      reason: 'Not running on a clean git client, aborting test.\n');
+
+  addTearDown(_resetGitClient);
+}
+
+Future<Null> _resetGitClient() async {
+  if (_gitIsClean()) return;
+
+  // Reset our state after each test, assuming we didn't abandon tests due
+  // to a non-pristine git environment.
+  Process.runSync('git', ['checkout', 'HEAD', '--', '.']);
+  // Delete the untracked files.
+  var gitStatus =
+      Process.runSync('git', ['status', '--porcelain', '.']).stdout as String;
+
+  Future<Null> nextBuild;
+  if (_process != null) nextBuild = nextSuccessfulBuild;
+  var untracked = gitStatus
+      .split('\n')
+      .where((line) => line.startsWith('??'))
+      .map((line) => line.split(' ')[1])
+      .map((path) {
+    var parts = p.split(path);
+    assert(parts[0] == 'e2e_example');
+    return p.joinAll(parts.skip(1));
+  });
+  for (var path in untracked) {
+    Process.runSync('rm', [path]);
```

Comment:
can we use `git clean` command for this?

Reply:
ah nice, didn't know about that one, done